DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
-    "                # Add batch to total_accuracy\n",
-    "                total_accuracy += result[4]\n",
-    "        \n",
-    "        # Print WER report\n",
-    "        if is_display_step:\n",
-    "            train_wer = calculate_and_print_wer_report( \\\n",
-    "                \"Training avg_cer=\" + \"{:.9f}\".format((total_accuracy / ceil(batches_per_device))), train_results)\n",
-    "            print\n",
+    "        print \"Training model...\"\n",
+    "        train_time = stopwatch(train_time)\n",
+    "        result = apply_batches(train_context, session, epoch=epoch, query_report=is_display_step)\n",
+    "        train_time = stopwatch(train_time)\n",
+    "\n",
+    "        result = print_report(\"Training\", result)\n",
+    "        # If there was a WER calculated, we keep it\n",
+    "        train_wer = result if result is not None else train_wer\n",
     "            \n",
     "        # Validation step\n",
     "        if is_validation_step:\n",
-    "            dev_wer = run_inference(session, \"Validation\", data_sets.dev)\n",
-    "\n",
-    "        # Checkpoint the model\n",
-    "        if is_checkpoint_step:\n",
+    "            print \"Hibernating training session into directory\", \"%s\" % checkpoint_dir\n",
     "            checkpoint_path = os.path.join(checkpoint_dir, 'model.ckpt')\n",
-    "            print \"Checkpointing in directory\", \"%s\" % checkpoint_dir\n",
-    "            saver.save(session, checkpoint_path, global_step=epoch)\n",
-    "            print\n",
+    "            # If the hibernation_path is set, the next epoch loop will load the model\n",
+    "            hibernation_path = stop_execution_context( \\\n",
+    "                train_context, session, checkpoint_path=checkpoint_path, global_step=epoch)\n",
     "            \n",
+    "            # Validating the moel in a fresh session\n",
+    "            print \"Validating model...\"\n",
+    "            result = run_set('dev', model_path=hibernation_path, query_report=(epoch == training_iters - 1))\n",
+    "            result = print_report(\"Validation\", result)\n",
+    "            # If there was a WER calculated, we keep it\n",
+    "            dev_wer = result if result is not None else dev_wer\n",
+    "\n",
     "        overall_time = stopwatch(overall_time)\n",
     "            \n",
     "        print \"FINISHED Epoch\", '%04d' % (epoch), \\\n",
-    "            \"   Overall time:\", format_duration(overall_time), \\\n",
-    "            \"  Training time:\", format_duration(training_time), \\\n",
-    "            \"   Average loss:\", \"{:.9f}\".format((loss / ceil(batches_per_device)))\n",
+    "            \"  Overall epoch time:\", format_duration(overall_time), \\\n",
+    "            \"  Training time:\", format_duration(train_time)\n",
     "        print\n",
-    "        \n",
+    "\n",
+    "    # If the last iteration step was no validation, we still have to save the model\n",
```

Comment:
So this would be much clearer if it read:
> If `hibernation_path` is `None`, then the model was not serialized and we need to do that here.
You could elaborate by then saying:
> This happens if validation was not performed on the preceding iteration.